### ISSUE TITLE
Improve error reporting: missing cast operator for subtype case

### ISSUE THREAD

**Issue by smoothdeveloper** [dotnet/fsharp#1128]
Related to #1103 #1127
## What

``` fsharp
let foo () : IDictionary<'a,'b> = Dictionary<'a,'b>()
```

`Error: This expression was expected to have type IDictionary<'a,'b> but here has type Dictionary<'a,'b>.`
## Why

New commers don't know about need for upcast, nor they know the idiomatic

``` fsharp
let foo () : IDictionary<'a,'b> = Dictionary<'a,'b>() :> _
```
## How

`Error: This expression was expected to have type IDictionary<'a,'b> but here has type Dictionary<'a,'b>, consider using cast operator :> _.`

**Comment by isaksky** [comment]
C# people can easily have problems here, because this works in CSharp:

![image](https://cloud.githubusercontent.com/assets/607986/14875359/46000396-0cc7-11e6-93ee-4c3cdc485981.png)

**Comment by forki** [comment]
I'm not sure if :>_ is really idiomatic here. I personally prefer to
suggest the concrete cast to IDictionary and think newcomers will
understand that better.

Anyways are you trying to implement this? If yes I can probably help if you
have specific questions.
On Apr 28, 2016 6:28 AM, "Isak Sky" <notifications@github.com> wrote:

C# people can easily have problems here, because this works in CSharp:

[image: image]
<https://cloud.githubusercontent.com/assets/607986/14875359/46000396-0cc7-11e6-93ee-4c3cdc485981.png>

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub
<https://github.com/Microsoft/visualfsharp/issues/1128#issuecomment-215307297>

**Comment by smoothdeveloper** [comment]
@forki, thanks a lot for your inspiring article / first code push on this effort for better error messages.

> I'm not sure if :>_ is really idiomatic here.

If you do C# or VB.NET no :smile: but I think it is idiomatic F#: usage of `_` to let compiler infer type / type parameters (especially in this case where it avoids repeating explicit type) is something I'd like to have seen / understood earlier.

> Anyways are you trying to implement this? If yes I can probably help if you
have specific questions.

Eventually, but I'm thinking of a easier one (#1129) if I'd ever get started contributing to the compiler.

**Comment by forki** [comment]
maybe we should suggest to use  `upcast`:

    let foo () : IDictionary<'a,'b> = upcast Dictionary<'a,'b>()

**Comment by isaacabraham** [comment]
I actually never knew about the ``:> _`` syntax until a few months ago - now I know it, I use it all the time - I think this is a situation were we could be a bit more proactive and give them this is as least as an option in the error message. I also think that we should add the "why" here again e.g. this works in C#, why not in F#. Maybe: -

```Error: This expression was expected to have type IDictionary<'a,'b> but here has type Dictionary<'a,'b>. Implicit upcasts are not permitted in this case. Consider either trying to use a flexible type, or explicitly upcasting this using the cast operator e.g. :> IDictionary<'a, 'b>, or through type inference e.g. :> _.```

This is probably a bit too wordy for some - but gets a bit more useful information in.

**Comment by forki** [comment]
mhm would work for me.

**Comment by baronfel** [comment]
Just a note that if/when [FS-1093](https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1093-additional-conversions.md) lands, many instances of this diagnostic should no longer be triggered (though new diagnostics might be triggered to let the user know the implicit upcast is occurring.)

**Comment by dsyme** [comment]
> Just a note that if/when FS-1093 lands, many instances of this diagnostic should no longer be triggered (though new diagnostics might be triggered to let the user know the implicit upcast is occurring.)

Closing this for this reason.  (There are likely other cases where we want a good diagnostic for subtypeing-related things but we should drive it with real examples.)
